### DISCUSSION TITLE
F0034 - Optional Procedure Arguments - Merged and Deleted

### DISCUSSION THREAD

**Discussion by adesutherland** [adesutherland/CREXX#133] [Progress Updates]
To try and make my explanation a bit more clear than on Thursday, I have created documentation and examples of how arguments are passed [here](https://github.com/adesutherland/CREXX/wiki/REXX-Assembler-(RXAS)#crexx-calling-convention)

**Comment by rvjansen** [comment]
I think I get it. I have corrected the prototypes in your global.rexx to work with the new convention. I am looking to correct the rxas bifs soon. One thing I am not clear about is what the consequences will be of the fact that now the number of arguments to a function call is possibly not 'correct' (if we left out optional arguments), but always corresponds to the maximum number in the prototype.

This is of course because I never have dealt with such a scenario; I would feel more secure with a 'max number' and an 'actual number' scheme, my feeling is that we will have to deal with things akin to SQL's 'null' and 'missing' confusion. (I always put in classifications of 'unknown' 'missing' 'not needed' etc - and skirt the whole issue).

For example, the length argument to substr() is optional, but what to put in the default? 0 seems straightforward, but 'missing' actually means until end-of-string here - if there is such a thing after 'start' gobbled them off. I can, of course, check the arguments and decrease the a0 for every register marked as 'unused', and keep the current logic mostly intact.

Also, what happens to user defined external functions and AOT compiling? Level C cannot require a prototype, and for functions that are not in the distribution in some way, we cannot know the signature.

So these are my initial questions (based on all of 20 minutes working with it, so I am sure more will be clear later).

**Reply by adesutherland** [reply]
Well for REXX Level B we will [eventually] have "..." for an arbitrary number of trailing parameters, and this would help.

For RXAS we can use the typeflag to detect an unspecified argument - so perhaps we can have a BIF like isargvalue(n) or some such name to get the status of the nth argument?

Clumsy but we can also do (never been tested!)
```
test: procedure = .int
 arg value = default_func()
 ...
 return

default_func: procedure = .int
  say "If I am called the argument to test() was not specified"
  return 0
```
But this is not very useful currently as there is no way to set a flag in test() from default_func()!!!

**Comment by rvjansen** [comment]
I did modify the two rxas versions of substr() already and they work fine when called from rexx now. So the excellent news is that we now can have compatible test suites for rexx and rxas versions of the bifs. Which I will work on this week.

**Comment by rvjansen** [comment]
Can it be that we are missing a SNES?

**Reply by rvjansen** [reply]
changed the 
```
\==
```
 to 
```
\= 
```
and we're fine

**Reply by adesutherland** [reply]
I don't really understand the meaning of  "strict" comparisons - at the moment they are treated the same (and/or some strict operators may be missing). Anyway an explanation of strict would be helpful.

**Reply by rvjansen** [reply]
Will look that up - it has got to do with spaces and insignificant zeroes I thought

**Reply by rvjansen** [reply]
"When comparing terms using these comparison operators, if both terms in the expression are numeric, REXX performs a numeric comparison. Otherwise, both terms are treated as character strings and REXX performs character comparison. (A number in REXX is a string that contains one or more decimal digits, an optional decimal point, and an optional leading sign character. The string can contain one or more leading or trailing blanks, and the sign character can be separated from the digits by one or more blanks.)

In a character comparison, leading and trailing blanks are ignored, and the shorter string is padded with blanks on the right. Character comparisons are case-sensitive.
In addition to these comparison operators, REXX provides a number of “strict” comparison operators that are mainly intended for use when comparing character strings. The strict comparison operators all have one of the characters defining the operator doubled, such as == (strictly equal).

When you use the == comparison operator (strictly equal), the two character strings being compared must be identical (character by character) and of the same length to be considered strictly equal. Leading and trailing blanks are significant.

Similarly, the strict comparison operators such as >> or << carry out a simple character-by-character comparison, with no padding of either of the strings being compared. The comparison of the two strings is from left to right. If one string is shorter than and is a leading substring of another, then it is smaller than (less than) the other. The strict comparison operators do not attempt to perform a numeric comparison on the two terms, and should not be used to compare numeric fields."

Note that in NetRexx, string comparison, when not strict, is case insensitive. Mike wanted to do this for REX but was talked out of it; for NetRexx, he implemented it as originally planned.

**Reply by adesutherland** [reply]
So we only do strict at the moment

I personally think that in cREXX Level B/G a non-strict string compare should also allow that "RENE" = "René" but for classic REXX Level C/D we should be case sensitive and so "René" /= "Rene". Level D needs to be backward compatible with level C

**Comment by rvjansen** [comment]
How does this work: I split out the functions from global.rexx in substr.rexx, raise.rexx and length.rexx. I test with a program called testsubstr.rexx. I need a prototype for substr in there, it is generated as global.substr. But each program file generates, resp. length.length, raise.raise and substr.substr ; if I edit those in the .rxas to global.* I am ok to test. Is there another way? I assume we need to specify the 'package' from the rexx files.

btw, we have three substr() now, and all have one or more errors. It is going to be a long day ;-)

**Reply by adesutherland** [reply]
At the moment the compiler just takes the file name as the "package" name. In the short term maybe you can append all the *.rexx files into a global.rexx?

When we have the class functionality the problem will be solved properly ...

Good luck with the debugging! I am shocked that there could be a bug in mine :-)

**Reply by adesutherland** [reply]
And - understanding your point - the prototype namespace is an issue - we need a quick fix for that

**Reply by rvjansen** [reply]
maybe just generate 'global' everywhere for the moment - that postpones the problem

**Reply by rvjansen** [reply]
concatenating the files does not help - it seems one can only have an 'options level' at the very beginning of the file

**Reply by rvjansen** [reply]
- or - there is something else that makes this segfault
```
Process 66477 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x22)
    frame #0: 0x0000000100050db0 rxc`step2a_walker(direction=in, node=0x00000001002163f0, payload=0x000000016fdfea40) at rxcpbval.c:377:33
   374
   375 	            /* Make a new symbol */
   376 	            symbol = sym_f(*current_scope, node);
-> 377 	            symbol->is_function = 1;
   378
   379 	            sym_adnd(symbol, node, 0, 1);
   380
Target 0: (rxc) stopped.
(lldb) exit
```

**Reply by rvjansen** [reply]
but I have abandoned that for the moment; 
```
/*
 * cREXX COMPILER VERSION : cREXX F0034
 * SOURCE                 : substr
 * BUILT                  : 2021-10-05 15:05:12
 */

.globals=0

raise() .expose=raise.raise

length() .expose=length.length

substr() .locals=15 .expose=global.substr
```
only one file needs editing.

**Comment by rvjansen** [comment]
actually, when we are going from a new assumption that called functions are coming out of their own file, only the called prototypes need to change, if they are prefixed with themselves. the problem is gone.

So testsubstr needs to expose substr.substr and substr needs to expose raise.raise and length.length  - If we do it like that, no editing is needed.

then this form can mean: current directory or library. a search directory or library can then be prepended.

**Reply by adesutherland** [reply]
Or shall I just make an even quicker short-term hack  and make everything global.xyx?

**Reply by rvjansen** [reply]
that might be the quickest short term remedy, but I would prefer the "called modules have their own name as prefix" - that way I might stumble upon other issues

**Reply by rvjansen** [reply]
but whatever is easier is best

**Reply by rvjansen** [reply]
this being level b, it would not be bad to enable an assembler escape to this expose statement; I understood earlier that at the moment, we only can specify rxvm machine instructions here

**Reply by adesutherland** [reply]
Not sure I understand - can you give me an example of what you mean?

**Reply by rvjansen** [reply]
Earlier I tried to use the 'assembler' statement of levelb to enter an 'expose x.y' into the .rxas, but that did not work - it needs an rxvm opcode, where expose is an assembler directive. But this is not actually needed if either everything will be global, *or* for the called functions an x.x is generated.

**Comment by rvjansen** [comment]
One thing is not clear to me. Where do the defaults of optional arguments go?
I have the following test exec:
```
/* rexx */
options levelb 
say '|'left('abc d',8)'|'
/* function prototype */
left: procedure = .string
arg string1 = .string, length = .int, pad = ' '
```

This generates:
```/*
 * cREXX COMPILER VERSION : cREXX I0134
 * SOURCE                 : ts_left
 * BUILT                  : 2021-10-07 00:30:14
 */

.globals=0

main() .locals=6 .expose=ts_left.main
   * Line 4: say '|'left('abc d',8)'|'
   * Line 4: left('abc d',8)
   load r1,3
   load r2,"abc d"
   settp r2,2
   load r3,8
   settp r4,2
   call r5,left(),r1
   concat r5,"|",r5
   concat r5,r5,"|"
   say r5
   * Line 4: 
   ret 0

left() .expose=left.left
```
So I see that the optional pad argument, with a default of ' ', has a type of 2, which is string. For me to successfully process the default pad in substr(string,start,[length],[pad]) I need to see that there are only three arguments and no pad, so I keep the default that is set in substr. So in fact, I need the type of r4 to be 0 (I set pad from substr) or r4 needs to contain the pad, ' ', but it does not. If I specify the pad in the call, all is well.

So to summarize, I think that either type = 2 and the pad string (default) needs to be in r4, or type = 0 and I infer the pad in substr.

**Reply by adesutherland** [reply]
So the type flags of r4 is set to 2 by the compiler.

Bit 0 is 0 which means the optional parameter is not present. You only need to look at bit 0 in substr, hence the 'and 1'

Bit 1 is set (2) ... this is the compiler being pedantic. Bit 1 set, means the caller does not require that the called procedure preserve the value of the argument because it is not linked to a variable.

Note: other values of the typeflag have not been defined (despite old documentation saying otherwise!)

Does this help?

**Reply by adesutherland** [reply]
Oh and you can see the compiler also set R2 type to 2. This is because the r2 is not a rexx variable. R2 is just being used temporarily to hold the string, so the compiler is giving the hint that it doesn't care if the called procedure alters the registers value.

Try

A = "blah"
B = substr(a, ...

To see the difference, in this case the type should be set to 0. The caller should not change the register value. 

The compiler uses this in functions to decide if it needs to do a string copy or not

**Reply by rvjansen** [reply]
The issue is that the pad, which should default to ' ', is not in the generated assembler.

**Reply by rvjansen** [reply]
But I will have a fresh look today, and will be back online about this.

**Reply by adesutherland** [reply]
Ah ... that is easy, the responsibility for setting the default value is in the called procedure, not in the code calling the procedure.

```
func() .locals=2 .expose=scratch.func
   * Line 15: x = "Adrian"
   brtpandt l28a,a1,1        * Check REGTP_VAL
   load r1,"Adrian"          * Set default Value <-****** HERE ******
   br l28d
l28a:
   brtpandt l28c,a1,2        * Check REGTP_NOTSYM (unless the default value has been used)
   scopy r1,a1               * Have to do a string copy
   br l28d
l28c:
   swap r1,a1                * Can just to a quick swap to get the register number right
l28d:
   ....
```

**Reply by rvjansen** [reply]
Makes sense. Let me try that.

**Reply by rvjansen** [reply]
works as designed!
